### PR TITLE
[Dev] Support LayerWiseDistributedOptimizer with torch_dist checkpoint format

### PR DESCRIPTION
**Pull request by BoxiangW**
[PR to main](https://github.com/NVIDIA/Megatron-LM/pull/2241)

Support variable DP size loading (DP-reshard). See https://wandb.ai/nvidia/muon-test/reports/Dist-ckpt-save-load-loss-verification--VmlldzoxNDkxODkyNw for a simple reference on loss curve report

It is loading at step 100 and the loss curve is overlapped with previous run

Screenshot 2025-10-31 at 11 34 50 AM

### CODE REVIEW

**Comment by skyw** [comment]
I remember the license check has been fixed. @BoxiangW could you update the branch and retry?

File: megatron/core/optimizer/optimizer.py [hunk 1]
```diff
@@ -788,6 +788,7 @@ def state_dict(self, is_loading: bool = False):
 
         state_dict = {}
         state_dict['optimizer'] = self.optimizer.state_dict()
+        # print(f"state_dict: {self.optimizer}, {state_dict['optimizer'].keys()}")
```
Comment (FDecaYed) [inline megatron/core/optimizer/optimizer.py]: accidental kept debug logging?

File: megatron/core/optimizer/muon.py [hunk 2]
```diff
@@ -280,22 +280,43 @@ def get_megatron_muon_optimizer(
...
             # creating master weight before layerwise sharding will lead to unnecessary master
-            # weight  so here we delay master weight creation into layer_wise unset config.bf16
+            # weight so here we delay master weight creation into layer_wise unset config.bf16
             # will also result in all optimizers below(adam) to also not be wrapped
             config.bf16 = False
             reset_config_bf16 = True
         else:
             # if not using layer_wise wrapper, just create master weight here is fine
-            optimizer = Float16OptimizerWithFloat16Params(optimizer, config, None, None)
+            optimizer = Float16OptimizerWithFloat16Params(optimizer, config, None, init_state_fn)
```
Comment (FDecaYed) [inline megatron/core/optimizer/muon.py]: `init_state_fn` not initialized here? should be muon_init_state_fn

**Review by FDecaYed — APPROVED** [review]
Some minor issues need fix but other than that code logic LGTM
